library/std/src/sys/unix/fs.rs

```diff
@@ -1683,8 +1683,21 @@ mod remove_dir_impl {
     fn remove_dir_all_recursive(parent_fd: Option<RawFd>, p: &Path) -> io::Result<()> {
         let pcstr = cstr(p)?;
 
-        // entry is expected to be a directory, open as such
-        let fd = openat_nofollow_dironly(parent_fd, &pcstr)?;
+        // try opening as directory
+        let fd = match openat_nofollow_dironly(parent_fd, &pcstr) {
+            Err(err) if err.raw_os_error() == Some(libc::ENOTDIR) => {
+                // not a directory
+                return match parent_fd {
+                    // return error if this is the root directory
+                    None => return Err(err),
```

Comment:
It confused me that we have `return` both on the `match` itself and in this arm. Personally, I'd take it off the `match` and add it into the other arm as well.

Reply:
I removed the `return` from the match arm, reordered them and adjusted the comments, if that is OK too.